zip-0302.rst

```diff
@@ -0,0 +1,76 @@
+::
+
+  ZIP: 302
+  Title: Standardized Memo Field Format
+  Owner: Jack Grigg <jack@electriccoin.co>
+  Original-Author: Jay Graber
+  Status: Draft
+  Category: Standards
+  Created: 2017-02-08
+  License: MIT
+
+
+Abstract
+========
+
+This ZIP describes a proposed specification for a standardized format for displaying content within the encrypted memo field of shielded transactions.
+
+Motivation
+==========
+
+A well-defined standard for formatting content within the encrypted memo field would help expand the use cases by providing a structure for different types of data. Users and third-party services could benefit from a standardized formatting convention that defines the type and length of the data contained within.
+
+A proposed common use case could be a standard encoding for sender’s preferred return address.
+
+Specification
+===============
+
+Section 5.5 of the Zcash protocol specification [#protocol-notept]_ defines three cases
+for the encoding of a memo field:
+
+* a UTF-8 human-readable string [#Unicode]_, padded by appending zero bytes; or
+* the byte ``0xF6`` followed by 511 ``0x00`` bytes, indicating "no memo"; or
+* any other sequence of 512 bytes starting with a byte value ``0xF5`` or greater
+  (which is therefore not a valid UTF-8 string), as specified in ZIP 302.
+
+This ZIP refines the specification of the third case.
+
++ If the first byte (byte 0)'s value is 0xF4 or smaller, then the reader should:
+     + strip any trailing zero bytes
+     + decode it as UTF-8 (replacing any incorrect UTF-8-encoded byte sequences with the replacement character U+FFFD), and display it to the user as a human-readable string.
```

Comment:
The trend in implementations is towards rejecting text memos that contain invalid UTF-8:
- The `Memo` struct in `zcash_primitives` has always [rejected invalid UTF-8](https://github.com/zcash/librustzcash/blob/8f087237123218b79b90537f08baf696bd295a09/zcash_primitives/src/note_encryption.rs#L112-L113).
- The `memoStr` field of the `z_viewtransaction` output is only returned by `zcashd` [if the memo is valid UTF-8](https://github.com/zcash/zcash/blob/35bb61be3900ff9d635de550c9efbe6b4ce96192/src/wallet/rpcwallet.cpp#L3748-L3750).
  - I made this decision consciously at the time, knowing it conflicted with this ZIP draft, but that relaxing the implemented conditions later was preferable to tightening them.
- I know @hdevalence wants `zebrad` to treat text memos with invalid UTF-8 as invalid, in order to have a typed memo representation that is 1:1 with the encrypted bytes.

I am leaning towards altering this ZIP to specify that invalid UTF-8 in text memos MUST be rejected, which implies that they would not be displayed to users, but the notes they are part of _would_ be treated as valid. Thoughts?

Currently in https://github.com/zcash/librustzcash/pull/177 I switch the `zcash_primitives` `Memo` struct to match this part of the ZIP draft, but I can easily revert that change.

Reply:
It seems to me that zcashd/zebrad shouldn't be internally opinionated about the contents of the memo field, as it's not a consensus-level concern, right? The interpretation of the memo field is entirely a wallet or other end-user level concern, so while one may want to use a newtype internally, for any consensus-relevant code that newtype should just be a wrapper around 512 uninterpreted bytes.

Reply:
@daira and I basically agree with @str4d here, consistency amongst implementations (not just on things pertaining to consensus) is a good thing.

Reply:
Sure; my point is merely that the contents of the memo field aren't consensus-critical - by analogy, the implementation of this likely shouldn't be part of zcash_primitives, but instead zcash_client_backend. I want to avoid a situation where some combination of 512 bytes could result in parse failure. (As I understand it, this is *not* intended to be a change to consensus rules.) 

From an implementation perspective, I think it's important to understand the round-trip serialization semantics of each of the parsing functions. This is, as I said, an implementation detail - if decoding a memo with a leading byte <= 0xF4 results in invalid UTF-8 characters, this should not result in a parse failure, but simply that the memo be interpreted as undifferentiated bytes.